src/Tools/Source/RunTests/HelixTestRunner.cs

```diff
@@ -0,0 +1,488 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Diagnostics.Contracts;
+using System.IO;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Security;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.TeamFoundation.TestManagement.WebApi;
+using Microsoft.VisualStudio.Services.Profile;
+using Newtonsoft.Json;
+
+namespace RunTests;
+
+public sealed class HelixWorkItem(
+    int id,
+    ImmutableArray<string> assemblyFilePaths,
+    ImmutableArray<string> testMethodNames)
+{
+    public string DisplayName { get; } = $"workitem_{id}";
+    public int Id { get; } = id;
+    public ImmutableArray<string> AssemblyFilePaths { get; } = assemblyFilePaths;
+    public ImmutableArray<string> TestMethodNames { get; } = testMethodNames;
+
+    public override string ToString() => DisplayName;
+}
+
+internal sealed class HelixTestRunner
+{
+    internal enum TestOS
+    {
+        Windows,
+        Linux,
+        Mac,
+    }
+
+    internal static async Task<int> RunAsync(Options options, ImmutableArray<AssemblyInfo> assemblies, CancellationToken cancellationToken)
+    {
+        Verify(options.UseHelix);
+        Verify(!options.IncludeHtml);
+        Verify(string.IsNullOrEmpty(options.TestFilter));
+        Verify(!string.IsNullOrEmpty(options.ArtifactsDirectory));
+        Verify(!string.IsNullOrEmpty(options.HelixQueueName));
+        Verify(!string.IsNullOrEmpty(options.Configuration));
+
+        // Currently, it's required for the client machine to use the same OS family as the target Helix queue.
+        // We could relax this and allow for example Linux clients to kick off Windows jobs, but we'd have to
+        // figure out solutions for issues such as creating file paths in the correct format for the target machine.
+        var testOS = (RuntimeInformation.IsOSPlatform(OSPlatform.Windows), RuntimeInformation.IsOSPlatform(OSPlatform.OSX)) switch
+        {
+            (true, _) => TestOS.Windows,
+            (_, true) => TestOS.Mac,
+            _ => TestOS.Linux
+        };
```

Comment:
nit: IMO, the following would be a preferable way to write this.

```suggestion
        var testOS = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? TestOS.Windows
            : RuntimeInformation.IsOSPlatform(OSPlatform.OSX) ? TestOS.Mac
            : TestOS.Linux;
```